File: test/language/enumeration_types/cpp/BitfieldConstEnumTest.cpp
#include <string>

#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"
#include "zserio/CppRuntimeException.h"

#include "enumeration_types/bitfield_const_enum/Color.h"

namespace enumeration_types
{
namespace bitfield_const_enum
{

class BitfieldConstEnumTest : public ::testing::Test
{
protected:
    static const size_t COLOR_BITSIZEOF;

    static const uint8_t NONE_VALUE;
    static const uint8_t RED_VALUE;
    static const uint8_t BLUE_VALUE;
    static const uint8_t GREEN_VALUE;
};

const size_t BitfieldConstEnumTest::COLOR_BITSIZEOF = 5;

const uint8_t BitfieldConstEnumTest::NONE_VALUE = 0;
const uint8_t BitfieldConstEnumTest::RED_VALUE = 2;
const uint8_t BitfieldConstEnumTest::BLUE_VALUE = 3;
const uint8_t BitfieldConstEnumTest::GREEN_VALUE = 7;

TEST_F(BitfieldConstEnumTest, EnumTraits)
{
    ASSERT_EQ("NONE", zserio::EnumTraits<Color>::names[0]);
    ASSERT_EQ("GREEN", zserio::EnumTraits<Color>::names[3]);
    ASSERT_EQ(4, zserio::EnumTraits<Color>::names.size());

    ASSERT_EQ(Color::RED, zserio::EnumTraits<Color>::values[1]);
    ASSERT_EQ(Color::BLUE, zserio::EnumTraits<Color>::values[2]);
    ASSERT_EQ(4, zserio::EnumTraits<Color>::values.size());
}

TEST_F(BitfieldConstEnumTest, enumToOrdinal)
{
    ASSERT_EQ(0, zserio::enumToOrdinal(Color::NONE));
    ASSERT_EQ(1, zserio::enumToOrdinal(Color::RED));
    ASSERT_EQ(2, zserio::enumToOrdinal(Color::BLUE));
    ASSERT_EQ(3, zserio::enumToOrdinal(Color::GREEN));
}

TEST_F(BitfieldConstEnumTest, valueToEnum)
{
    ASSERT_EQ(Color::NONE, zserio::valueToEnum<Color>(NONE_VALUE));
    ASSERT_EQ(Color::RED, zserio::valueToEnum<Color>(RED_VALUE));
    ASSERT_EQ(Color::BLUE, zserio::valueToEnum<Color>(BLUE_VALUE));
    ASSERT_EQ(Color::GREEN, zserio::valueToEnum<Color>(GREEN_VALUE));
}

TEST_F(BitfieldConstEnumTest, valueToEnumFailure)
{
    ASSERT_THROW(zserio::valueToEnum<Color>(1), zserio::CppRuntimeException);
}

TEST_F(BitfieldConstEnumTest, bitSizeOf)
{
    ASSERT_TRUE(zserio::bitSizeOf(Color::NONE) == COLOR_BITSIZEOF);
}

TEST_F(BitfieldConstEnumTest, initializeOffsets)
{
    const size_t bitPosition = 1;
    ASSERT_TRUE(zserio::initializeOffsets(bitPosition, Color::NONE) == bitPosition + COLOR_BITSIZEOF);
}

TEST_F(BitfieldConstEnumTest, read)
{
    zserio::BitStreamWriter writer;
    writer.writeBits(static_cast<uint32_t>(Color::RED), COLOR_BITSIZEOF);
    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);

    Color color(zserio::read<Color>(reader));
    ASSERT_EQ(RED_VALUE, zserio::enumToValue(color));
}

TEST_F(BitfieldConstEnumTest, write)
{
    const Color color(Color::BLUE);
    zserio::BitStreamWriter writer;
    zserio::write(writer, color);

    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);
    ASSERT_EQ(BLUE_VALUE, reader.readBits(COLOR_BITSIZEOF));
}

} // namespace bitfield_const_enum
} // namespace enumeration_types

File: test/language/enumeration_types/python/BitfieldConstEnumTest.py
import unittest

import zserio

from testutils import getZserioApi

class BitfieldEnumTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "enumeration_types.zs",
                               extraArgs=["-withTypeInfoCode"]).bitfield_const_enum

    def testValues(self):
        self.assertEqual(NONE_VALUE, self.api.Color.NONE.value)
        self.assertEqual(RED_VALUE, self.api.Color.RED.value)
        self.assertEqual(BLUE_VALUE, self.api.Color.BLUE.value)
        self.assertEqual(GREEN_VALUE, self.api.Color.GREEN.value)

    def testFromString(self):
        self.assertEqual(self.api.Color.from_name("NONE"), self.api.Color.NONE)
        self.assertEqual(self.api.Color.from_name("RED"), self.api.Color.RED)
        self.assertEqual(self.api.Color.from_name("BLUE"), self.api.Color.BLUE)
        self.assertEqual(self.api.Color.from_name("GREEN"), self.api.Color.GREEN)
        with self.assertRaises(zserio.PythonRuntimeException):
            self.api.Color.from_name("NONEXISTING")

    def testFromReader(self):
        writer = zserio.BitStreamWriter()
        writer.write_bits(self.api.Color.GREEN.value, COLOR_BITSIZEOF)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        color = self.api.Color.from_reader(reader)
        self.assertEqual(GREEN_VALUE, color.value)

    def testCalcHashCode(self):
        # use hardcoded values to check that the hash code is stable
        self.assertEqual(1702, zserio.hashcode.calc_hashcode_object(zserio.hashcode.HASH_SEED,
                                                                    self.api.Color.NONE))
        self.assertEqual(1704, zserio.hashcode.calc_hashcode_object(zserio.hashcode.HASH_SEED,
                                                                    self.api.Color.RED))
        self.assertEqual(1705, zserio.hashcode.calc_hashcode_object(zserio.hashcode.HASH_SEED,
                                                                    self.api.Color.BLUE))
        self.assertEqual(1709, zserio.hashcode.calc_hashcode_object(zserio.hashcode.HASH_SEED,
                                                                    self.api.Color.GREEN))

    def testBitSizeOf(self):
        self.assertEqual(COLOR_BITSIZEOF, self.api.Color.NONE.bitsizeof())
        self.assertEqual(COLOR_BITSIZEOF, self.api.Color.RED.bitsizeof())
        self.assertEqual(COLOR_BITSIZEOF, self.api.Color.BLUE.bitsizeof())
        self.assertEqual(COLOR_BITSIZEOF, self.api.Color.GREEN.bitsizeof())

    def testInitializeOffsets(self):
        self.assertEqual(COLOR_BITSIZEOF, self.api.Color.NONE.initialize_offsets(0))
        self.assertEqual(COLOR_BITSIZEOF + 1, self.api.Color.RED.initialize_offsets(1))
        self.assertEqual(COLOR_BITSIZEOF + 2, self.api.Color.BLUE.initialize_offsets(2))
        self.assertEqual(COLOR_BITSIZEOF + 3, self.api.Color.GREEN.initialize_offsets(3))

    def testWrite(self):
        writer = zserio.BitStreamWriter()
        self.api.Color.RED.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        self.assertEqual(RED_VALUE, reader.read_bits(COLOR_BITSIZEOF))

COLOR_BITSIZEOF = 5

NONE_VALUE = 0
RED_VALUE = 2
BLUE_VALUE = 3
GREEN_VALUE = 7
